Source/DafnyTestGeneration/TestMethod.cs

```diff
@@ -64,12 +76,39 @@ public class TestMethod {
     /// Returns a string that contains all the {:synthesize} annotated methods
     /// necessary to compile the tests
     /// </summary>
-    public static string EmitSynthesizeMethods() {
+    public static string EmitSynthesizeMethods(DafnyInfo dafnyInfo) {
       var result = "";
-      foreach (var typ in TypesToSynthesize) {
+      foreach (var typ in TypesToSynthesize.Keys) {
         var methodName = GetSynthesizeMethodName(typ);
-        result += $"\nmethod {{:synthesize}} {methodName}() " +
-                  $"returns (o:{typ}) ensures fresh(o)";
+        var returnName = "o";
+        if (dafnyInfo.IsTrait(TypesToSynthesize[typ])) {
+          var types = dafnyInfo.GetTypesForTrait(TypesToSynthesize[typ]);
+          int id = 0;
+          var typeToNameDict = new Dictionary<string, string>();
+          foreach (var resultTyp in types) {
+            typeToNameDict[resultTyp.ToString()] = "o_res_" + id++;
+          }
+          typeToNameDict[typ] = "o";
+          result += $"\nmethod {{:synthesize}} {methodName}(" +
```

Comment:
It would be lovely to eventually use a pretty-printer for Dafny source, and some nice combinators to construct common patters, and emit this from an AST. Maybe it could even type check the result! But that's future work, I think.

Reply:
Yes, outputting an AST rather than a string is high on my priority list because it would also allows people to optionally modify the AST before printing to file.